### PR TITLE
ReplaceJumpWithRet

### PR DESCRIPTION
**Pull request by Hecate2**
If a JMP or JMP_L jumps to a RET, replace the JMP with RET

### CODE REVIEW

**Comment by Hecate2** [comment]
This is the first time I add new instructions to the optimized contract, and the added RET is not binded to source codes with debuginfo. For now this may not be a big problem, but in the future we may need to bind new instructions to source codes.

**Comment by shargon** [comment]
This make harder the method detection, currently we only have one ret for each method

**Comment by Jim8y** [comment]
> This make harder the method detection, currently we only have one ret for each method

do we have limitation on using ret? i dont know that.

**Comment by Hecate2** [comment]
> This make harder the method detection, currently we only have one ret for each method

Are we now really restricting the count of RET and detecting methods using the only RET? I can write a robust method detection if needed. I can also abandon this PR and (slowly) write something else to ensure only a single RET exists in a method.

**Comment by roman-khimov** [comment]
Go compiler can generate as many RETs for a method as needed and typical Go style even encourages that. Method boundaries are usually tracked with manifests, it's just more reliable.

**Review by Jim8y — APPROVED** [review]

File: src/Neo.Compiler.CSharp/Optimizer/Strategies/Reachability.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ namespace Neo.Optimizer
 {
     static class Reachability
     {
-        [Strategy(Priority = int.MaxValue)]
+        [Strategy(Priority = int.MaxValue - 16)]
```
[the remark this answers is not in the document]
Comment (Hecate2) [reply src/Neo.Compiler.CSharp/Optimizer/Strategies/Reachability.cs:25]: It means nothing. Just reserving spaces for other optimizers. We are not even using the `Strategy` attribute now.

**Comment by Jim8y** [comment]
> o

@shargon Hey shargon, any further detail about this?

**Comment by shargon** [comment]
It complicate the reverse engineering but maybe we should take care more avoid optimization instead of reversing

**Pull request merged**